### PR TITLE
XML parser - attempt 1

### PR DESCRIPTION
**Pull request by nessime7**

### CODE REVIEW

File: src/main/java/XmlParser.java [hunk 1]
```diff
@@ -1,4 +1,41 @@
...
+import java.io.IOException;
+
 public class XmlParser {
+    // proszę o sprawdzenie
     public static void main(String[] args) {
+        String file = args[0];
+        parser(file);
+    }
+
+    public static String parser(String file) {
```
Comment (albert-podraza) [inline src/main/java/XmlParser.java]: Please create dedicated class for parsing logic and use it

File: src/main/java/XmlParser.java [hunk 2]
```diff
@@ -1,4 +1,41 @@
...
+                if (node.getNodeType() == Node.ELEMENT_NODE) {
+                    Element element = (Element) node;
+                    String name = element.getElementsByTagName("name").item(0).getTextContent();
+                    String email = element.getElementsByTagName("email").item(0).getTextContent();
+                    String userName = element.getElementsByTagName("username").item(0).getTextContent();
+                    System.out.println("name: " + name + " email: " + email + " userName: " + userName);
+                }
+            }
+        } catch (SAXException | IOException | ParserConfigurationException e) {
+            System.err.println("Error parsing XML: " + e.getMessage());
```
Comment (albert-podraza) [inline src/main/java/XmlParser.java:29]: Why did you use err class object instaead of out ?

File: src/main/java/XmlParser.java [hunk 3]
```diff
@@ -1,4 +1,41 @@
...
 public class XmlParser {
+    // proszę o sprawdzenie
     public static void main(String[] args) {
+        String file = args[0];
+        parser(file);
+    }
+
+    public static String parser(String file) {
+        try {
+            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
```
Comment (albert-podraza) [inline src/main/java/XmlParser.java]: please use final var wherever it's possible 
https://developers.redhat.com/blog/2018/05/25/simplify-local-variable-type-definition-using-the-java-10-var-keyword

File: src/main/java/XmlParser.java [hunk 4]
```diff
@@ -1,4 +1,41 @@
...
+            Document document = builder.parse(file);
+            NodeList nodeList = document.getDocumentElement().getChildNodes();
+            for (int i = 0; i < nodeList.getLength(); i++) {
+                Node node = nodeList.item(i);
+                if (node.getNodeType() == Node.ELEMENT_NODE) {
+                    Element element = (Element) node;
+                    String name = element.getElementsByTagName("name").item(0).getTextContent();
+                    String email = element.getElementsByTagName("email").item(0).getTextContent();
+                    String userName = element.getElementsByTagName("username").item(0).getTextContent();
+                    System.out.println("name: " + name + " email: " + email + " userName: " + userName);
```
Comment (albert-podraza) [inline src/main/java/XmlParser.java]: please create dedicated class for printing output and use it

File: src/main/java/XmlParser.java (hunk 1 above)
Comment (albert-podraza) [inline src/main/java/XmlParser.java]: Please use OOP principles, not procedural paradigm
https://medium.com/@LiliOuakninFelsen/functional-vs-object-oriented-vs-procedural-programming-a3d4585557f3

File: src/test/java/XmlParserTest.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class XmlParserTest {
+
+    @Test
+    void should_should_return_users_from_users_xml() {
+        // given
+        String file = "users.xml";
```
Comment (albert-podraza) [inline src/test/java/XmlParserTest.java]: To make the test successful, most likely this file should be in test resources, but you need to check it

File: src/test/java/XmlParserTest.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+        // when
+        var result = XmlParser.parser(file);
+
+        // then
+        assertEquals(result,"name: Kalle Anka email: donald@email.dt userName: donaldd\n" +
```
Comment (albert-podraza) [inline src/test/java/XmlParserTest.java]: When you split your class to 3 smaller ones, only parsing class should be tested, it should ease testing

File: src/main/java/XmlParser.java (hunk 1 above)
Comment (albert-podraza) [reply src/main/java/XmlParser.java]: Also I encourage to read https://www.jakowicz.com/what-is-apie/ in regards to the above

**Review by albert-podraza — COMMENTED** [review]
Commit name could be "review fixes"

**Review by albert-podraza — COMMENTED** [review]
You shouldn't include .gradle, .idea, build directories to pull request, please read about https://www.atlassian.com/pl/git/tutorials/saving-changes/gitignore and remove them

**Comment by albert-podraza** [comment]
You didn't address or partially addressed my comments, please have a look at them again and address them.

File: src/main/java/Main.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+public class Main {
+
+    private static XmlParser xmlParser;
+
+    public static void main(String[] args) {
+        String file = args[0];
+        xmlParser.parser(file);
```
Comment (albert-podraza) [inline src/main/java/Main.java]: I'm pretty sure it's not working, it's not good you even didn't check your program if it runs after the changes

Comment (nessime7) [reply src/main/java/Main.java]: It's working from Jar.

**Pull request closed without merging**